add a layout debug mode that sets background colors

aerickson opened the pull request:
Makes debugging Toga's layout system easier by visually indicating container boundaries. See https://github.com/beeware/toga/issues/2856.

features:
- randomized background color (using colors optimized for color-blindness) for boxes
- enabled via TOGA_DEBUG_LAYOUT env var

usage:
```bash
TOGA_DEBUG_LAYOUT=1 briefcase dev
```

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

aerickson commented:
Throwing this up for initial feedback, will complete docs and tests once things are looking good.

HalfWhitt commented on core/src/toga/widgets/base.py:
```diff
@@ -33,6 +54,20 @@ def __init__(
             will be applied to the widget.
         """
+        # if the object has _USE_DEBUG_BACKGROUND=True and layout debug mode is on, change bg color
+        if hasattr(self, "_USE_DEBUG_BACKGROUND") and self._USE_DEBUG_BACKGROUND:
```
Hope I'm not butting in here, but a handy thing you may not be aware of: [`getattr()`](https://docs.python.org/3/library/functions.html#getattr) can take an optional third parameter as a default to supply if the attribute is missing. Dictionaries have a roughly analogous [`get()`](https://docs.python.org/3/library/stdtypes.html#dict.get) method, which returns `None` or a user-supplied default if the key is missing.

So lines 58 and 59 could do the same thing and be written as:

```python
if getattr(self, "_USE_DEBUG_BACKGROUND", False):
    if environ.get("TOGA_DEBUG_LAYOUT") == "1":
```

freakboy3742 commented on changes/2953.feature.rst:
```diff
@@ -0,0 +1 @@
+Added a layout debugging mode that sets containers to random background colors.
```
The change note is the draft of the release note for the feature, rather than being a commit summary; so - it should be worded in terms of "exiting new feature you can now use!!1!!BBQ!":
```suggestion
Toga now has a layout debugging mode. If you set ``TOGA_DEBUG_LAYOUT=1`` in your app's runtime environment, widgets will be rendered with different background colors, making it easier to identify how space is being allocated by Toga's layout algorithm.
```

freakboy3742 commented on core/src/toga/widgets/base.py:
```diff
@@ -33,6 +54,20 @@ def __init__(
+        # if the object has _USE_DEBUG_BACKGROUND=True and layout debug mode is on, change bg color
+        if hasattr(self, "_USE_DEBUG_BACKGROUND") and self._USE_DEBUG_BACKGROUND:
+            if "TOGA_DEBUG_LAYOUT" in environ and environ["TOGA_DEBUG_LAYOUT"] == "1":
+                if Widget._debug_color_index == len(debug_background_palette) - 1:
+                    Widget._debug_color_index = 0
+                else:
+                    Widget._debug_color_index += 1
+                style = style if style else Pack()
+                style.background_color = debug_background_palette[
+                    Widget._debug_color_index
+                ]
```
```suggestion
                Widget._debug_color_index += 1
                style = style if style else Pack()
                style.background_color = debug_background_palette[
                    Widget._debug_color_index % len(debug_background_palette)
                ]
```

HalfWhitt replied:
I don't *think* it would make a difference here, since one blank Pack should be the same as any other, but checking `style is not None` would be more consistent with other things that will interact with future versions of Travertino...

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,57 @@
+def test_button_debug_background():
+    """A Button in layout debug mode has a default background."""
+    # Enable layout debug mode
+    environ["TOGA_DEBUG_LAYOUT"] = "1"
```
This modifies the global environment, so it will leak into external tests (which means this test can impact on other tests). For something like this, it's a good idea to use [`monkeypatch.setenv`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.MonkeyPatch.setenv) - that will automatically reset the environment variable when the test finishes.

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,57 @@
+
+    # need enough for coverage of palette array index rollover
+    box = toga.Box()
+    toga.Box()
```
```
for i in range(0, 10):
...
```

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,57 @@
+
+
+# test that a non-container-like widget in layout debug mode has a default background
+def test_button_debug_background():
```
Good idea testing a widget other than Box; - but also worth adding at test for (a) a container widget, and (b) a widget where the debug background *won't* be applied.

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,57 @@
+
+    # assert that the bg is not default/white
+    assert hasattr(box.style, "background_color")
+    assert not box.style.background_color == rgb(0, 0, 0)
```
"Not white" can cover a multitude of sins :-) 

A better test would be to evaluate the *specific* colors for *all* the widgets created in a specific order. The order of allocation should be predictable, once you've found the color of the first box.

freakboy3742 reviewed (changes requested):
I know this is still marked as a draft, but I couldn't help myself providing some initial feedback. tl;dr - this is all looking very promising; with a couple of fairly minor tweaks, it should be pretty close to merge-ready.

aerickson commented:
> I know this is still marked as a draft, but I couldn't help myself providing some initial feedback. tl;dr - this is all looking very promising; with a couple of fairly minor tweaks, it should be pretty close to merge-ready.

Awesome. :) 

I think I've addressed all of the comments. Thank you both.

freakboy3742 commented:
@aerickson Happy to give this another review; but it's currently showing a merge conflict that needs to be resolved. Also, you can flick this over to "ready for review" if you think it's ready.

freakboy3742 commented:
The CI failure isn't anything to be concerned about - it's a recently introduced intermittent failure (#3016).

aerickson commented:
> @aerickson Happy to give this another review; but it's currently showing a merge conflict that needs to be resolved. Also, you can flick this over to "ready for review" if you think it's ready.

Changed to 'ready for review'. Rebased.

> The CI failure isn't anything to be concerned about - it's a recently introduced intermittent failure (#3016).

OK, thanks. :)

aerickson commented:
Where would you like this documented? 

https://toga.readthedocs.io/en/stable/how-to/topics/layout.html (aka docs/how-to/topics/layout.rst) maybe?

freakboy3742 commented on core/src/toga/widgets/base.py:
```diff
@@ -17,10 +19,29 @@
+    "#e5e4af",  # light cream
+    "#bde2dc",  # soft turquoise
+]
+shuffle(debug_background_palette)
```
Why programmatically shuffle? We definitely need an order which cycles good contrast (i.e., we don't have "light blue" next to "very light blue"), but that order doesn't need to be random.

Plus, randomisation takes more time at startup, and a randomised order could be jarring for as successive runs won't give the same output.

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+
+
+# test that a container-like widget in normal mode has a default background
+def test_box_no_background_in_normal_mode():
+    """A Box has no default background."""
+    # Disable layout debug mode
+    with MonkeyPatch.context() as mp:
+        mp.setenv("TOGA_DEBUG_LAYOUT", "0")
```
Monkeypatch can be used as a fixture:
```suggestion
def test_box_no_background_in_normal_mode(monkeypatch):
    """A Box has no default background."""
    # Disable layout debug mode
    monkeypatch.setenv("TOGA_DEBUG_LAYOUT", "0")
```

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
We try to keep the test structure matching the widget layout; or, if things are getting complex, matching the feature by the name used in the code. In this case, if there's too many tests to add to `widgets/test_base.py`, this should probably be `widgets/test_debug_layout.py` (I think the latter makes sense here)

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+        mp.setenv("TOGA_DEBUG_LAYOUT", "0")
+        box = toga.Box()
+        # assert that the bg is default
+        assert hasattr(box.style, "background_color")
```
This assertion essentially *can't* fail. Toga's style framework will ensure that `background_color` exists as an attribute for read.

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+        box = toga.Box()
+        # assert that the bg is default
+        assert hasattr(box.style, "background_color")
+        assert not box.style.background_color
```
A check of an actual value, rather than "falseness" would be preferable here - that avoids any odd situations where something like `rgb(0,0,0)` evaluates as "false" (I don't think it does, but just in case)

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+        mp.setenv("TOGA_DEBUG_LAYOUT", "1")
+        button = toga.Button()
+        # assert that the bg is default
+        assert hasattr(button.style, "background_color")
+        assert not button.style.background_color
```
As above:
* monkeypatch can be used as a fixture.
* the `hasattr` won't be asserting anything
* A positive value check would be preferable.

The same pattern applies to the subseqeunt tests as well.

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+        assert not button.style.background_color
+
+
+# test that a label in layout debug mode has a default background
```
Why test label *and* button? Or, if it's important to check every widget (arguable, it is), is there any reason we shouldn't encode this as a test in `properties.py` like `test_enabled`/`test_enable_noop` - that is, a test that can be applied to every widget, with a "uses debug" and "doesn't use debug" variant?

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+
+        for counter, box in enumerate(boxes, start=1):
+            index = counter % debug_bg_palette_length
+            print(counter)
```
```suggestion
```

freakboy3742 commented on core/src/toga/widgets/base.py:
```diff
@@ -17,10 +19,29 @@
 
 
+# based on colors from https://davidmathlogic.com/colorblind
+debug_background_palette = [
```
If this is a list of constants, it should be treated as a constant by naming convention:
```suggestion
DEBUG_BACKGROUND_PALETTE = [
```

freakboy3742 commented on core/tests/widgets/test_container_debug_mode.py:
```diff
@@ -0,0 +1,76 @@
+            print(counter)
+            assert hasattr(box.style, "background_color")
+            assert box.style.background_color == color(
+                toga.widgets.base.debug_background_palette[index]
```
Unless I'm missing something, this test doesn't account for the state of `_debug_color_index` at the start of the test. it essentially *must* be the first (and only?) "background modifying" test in the suite; if any other test enables debug, this test will fail because the colors won't align.

This means either the test needs to monkeypatch `_debug_color_index`, or the test needs to take the initial state of `_debug_color_index` into consideration.

freakboy3742 reviewed (changes requested):
This is definitely heading in the right direction - a couple of suggestions inline, mostly about testing. The most significant of those suggestions is possibly switch to making this a parameterised test that can be evaluated on *every* widget, which would then give us a confirmation that we've got `_USE_DEBUG_BACKGROUND` set correctly for each widget.

Regarding documentation - I think this should be the start of a new HOWTO topic guide on "debugging your app"; that document will only have one section for now (debugging layout); but it gives us a place to put other debugging hints (e.g., how to use pdb et al). There should also be a cross reference to this new section in the documentation on Pack, so that someone trying to understand Pack is made aware that the debugging option exists.

HalfWhitt commented:
Superseded by #3600
